### PR TITLE
CLI usage of new `as_feature` function of report/indicator classes

### LINKED ISSUE #12: CLI: GeoJSON input should be agnostic to whether a Feature or FeatureCollection get passed in

### LINKED ISSUE #64: Missing indicator description in CLI output
Calculating an indicator or report, the metadata contains an attribute `description `which describes the indicator, and the results contain an attribute `description` as well, which describes the result. Due to the matching name, the result-description overwrites the indicator-description.

On the webpage, the metadata-description is called `Report description`/`Indicator description`. I would suggest to use the same naming in the metadata class

### LINKED ISSUE #67: CLI: Should output be a new GeoJSON or should the input GeoJSON be updated?
Should output be a new GeoJSON or should the input GeoJSON be updated? (Preserve properties of input GeoJSON Object)

### LINKED ISSUE #84: More granular output in GeoJSON when creating a report through the CLI
It would be great if the output for a report could contain more detailed information on each of the indicators. Currently (see image attached), the GeoJSON output includes the overall label (red/yellow/green) and an aggregated value. Would it be possible to also include the value associated with each indicator in the report? 

Capture

### PR DESCRIPTION
**Pull request by matthiasschaub**
CLI usage of new `as_feature` function of report/indicator classes

### Description
Please add a clear and concise description of what your PR solves.

### Corresponding issue
Closes #84 
Closes #67
Closes #64 
Closes #12

### Checklist
- [x] I have updated my branch to ~`main`~ `geo-interface-cli` (e.g. through `git rebase main`)
- [x] My code follows the [style guide](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#style-guide) and was checked with [pre-commit](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#pre-commit) before committing
- [x] I have commented my code
- [x] I have added sufficient unit and integration [tests](https://github.com/GIScience/ohsome-quality-analyst/blob/main/docs/development_setup.md#tests)
- [x] I have updated the [CHANGELOG.md](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CHANGELOG.md)

Please check all finished tasks. If some tasks do not apply to your PR, please cross their text out (by using `~...~`) and remove their checkboxes.

### CODE REVIEW

File: workers/ohsome_quality_analyst/cli.py [hunk 1]
```diff
@@ -214,18 +212,13 @@ def create_report(
...
-        else:
-            click.echo(report.metadata)
-            click.echo(report.result)
+        geojson_object = report.as_feature()
+    if outfile:
+        write_geojson(outfile, geojson_object)
+    else:
+        click.echo(
+            geojson.dumps(geojson_object, default=datetime_to_isostring_timestamp)
+        )
```
Comment (joker234) [inline workers/ohsome_quality_analyst/cli.py:221]: is it good to have something like this dependent on the parameter `outfile`? Wouldn't it be better if I could have both outputs at the same time?

File: workers/ohsome_quality_analyst/cli.py [hunk 2]
```diff
@@ -105,20 +106,20 @@ def create_indicator(
 ):
     """Create an Indicator.
 
-    Write a GeoJSON with the result to disk if an outfile is specified.
+    Output is a GeoJSON Feature or FeatureCollection with the indicator results.
+    Write to disk if the `outfile` option is specified.
```
Comment (joker234) [inline workers/ohsome_quality_analyst/cli.py:110]: see other comment about the `outfile` parameter. If the code stays at it is, it should be stated that the output **only** goes to disk, if `outfile` is specified.

File: workers/tests/integrationtests/test_cli_file_io.py [hunk 3]
```diff
@@ -31,24 +31,24 @@ def test_create_indicator_fcollection_outfile(self):
...
+                "--infile",
+                infile,
+                "--outfile",
+                outfile,
+            ],
+        )
+        self.assertEqual(result.exit_code, 0)
+        with open(outfile, "r") as testobject:
+            testobject = geojson.load(testobject)
+        self.assertTrue(testobject.is_valid)
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/tests/integrationtests/test_cli_file_io.py:51]: I was inattentive. You are right. I will reverse those changes.

File: workers/ohsome_quality_analyst/utils/helper.py [hunk 4]
```diff
@@ -71,39 +73,31 @@ def datetime_to_isostring_timestamp(time: datetime) -> str:
...
-    """loads the input file as feature collection which will be used as input for
-    indicator or report calculations"""
-    infile = pathlib.Path(infilepath)
-    with open(infile, "r") as file:
-        feature_collection = geojson.load(file)
-    if feature_collection.is_valid is False:
-        raise ValueError("Input geometry is not valid")
-    return feature_collection
-
+def write_geojson(outfile, geojson_object):
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/utils/helper.py:76]: Added type hints

File: workers/ohsome_quality_analyst/cli_opts.py [hunk 5]
```diff
@@ -34,14 +34,20 @@
 
 infile = click.option(
     "--infile",
-    help="GeoJSON file for your area of interest.",
+    help=(
+        "Path to a GeoJSON file. "
+        + "Geometry has to be of type Polygon or MultiPolygon."
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/cli_opts.py:39]: Removed the plus

File: workers/ohsome_quality_analyst/cli_opts.py [hunk 6]
```diff
@@ -34,14 +34,20 @@
...
     type=click.Path(resolve_path=True),
     default=None,
 )
 
 outfile = click.option(
     "--outfile",
-    help="GeoJSON file to be written with appended metadata and results.",
+    help=(
+        "Path to a GeoJSON file. "
+        + "Path and file will be created if it does not exists."
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/cli_opts.py:49]: Removed the plus.

File: workers/ohsome_quality_analyst/cli.py (hunk 1 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/cli.py:221]: Then the whole GeoJSON would be printed to stdout all the time.
We could also printout only the results not the geometry for example.

Not sure what the best approache here is.

What does the user expect? What is the most helpful data printed to stdout?

File: CHANGELOG.md [hunk 7]
```diff
@@ -28,6 +28,7 @@
 - Redefine OQT regions ([#26])
 - Use ohsome API endpoint `/contributions/latest/count` for Last Edit indicator ([#68])
 - Implement `as_feature` function for indicator and report classes ([#86])
+- CLI usage of new `as_feature` function of report and indicator classes ([#87])
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply CHANGELOG.md]: I made the description more clear.

File: workers/ohsome_quality_analyst/cli.py (hunk 1 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/cli.py]: I changed the behavior to always print to stdout and made the help text more explicit.

Regarding the output of a not so readable GeoJSON to stdout see issue: https://github.com/GIScience/ohsome-quality-analyst/issues/89

File: workers/ohsome_quality_analyst/cli.py (hunk 2 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/cli.py]: See answer to above comment.

File: workers/ohsome_quality_analyst/utils/helper.py [hunk 8]
```diff
@@ -71,39 +73,31 @@ def datetime_to_isostring_timestamp(time: datetime) -> str:
...
-    feature["properties"].update(vars(indicator.result).copy())
-    return feature
-
-
-def update_features_report(feature, report):
-    """adds the report metadata and result to the properties of a feature"""
-    feature["properties"].update(vars(report.metadata))
-    feature["properties"].update(vars(report.result))
-    return feature
+def loads_geojson(bpolys: str) -> Generator[Feature, None, None]:
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/utils/helper.py:90]: Implicit testing through tests for cli.py was present.

But I added some unittests for this function!

File: workers/ohsome_quality_analyst/cli.py (hunk 1 above)
Comment (joker234) [reply workers/ohsome_quality_analyst/cli.py]: thanks :))

File: workers/ohsome_quality_analyst/cli_opts.py [hunk 9]
```diff
@@ -35,8 +35,7 @@
 infile = click.option(
     "--infile",
     help=(
-        "Path to a GeoJSON file. "
-        + "Geometry has to be of type Polygon or MultiPolygon."
+        "Path to a GeoJSON file. " "Geometry has to be of type Polygon or MultiPolygon."
```
Comment (joker234) [inline workers/ohsome_quality_analyst/cli_opts.py:38]:
```suggestion
        "Path to a GeoJSON file. Geometry has to be of type Polygon or MultiPolygon."
```

**Review by joker234 — COMMENTED** [review]
Only a very small thing you could change. Thanks for all the tests :bow:

**Comment by matthiasschaub** [comment]
> Only a very small thing you could change. Thanks for all the tests bow

Thanks for the review :)

**Pull request merged**